### PR TITLE
MacOS packaging build with Gitlab CI

### PR DESCRIPTION
**Pull request by maximedenes**
This PR is similar to #7344, but for macOS.

### CODE REVIEW

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml]: ?

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml]: This change doesn't do anything, try sticking a `printenv` to see what variables are defined.

**Comment by Zimmi48** [comment]
Same remark as for Windows. In the long run it would probably be profitable to use Docker here as well. I'm saying this right now because you might consider whether it could be worth to do it using Docker from the start.

**Comment by maximedenes** [comment]
I'm already late by two weeks on this work, so I won't be able to explore the use of docker for now. But I agree it's probably the way to go.

**Comment by Zimmi48** [comment]
Yah, if you feel that this is the easiest way to proceed for now, I'm fine with this as a first step.

**Comment by Zimmi48** [comment]
TODO as part of this PR: remove the Travis macOS packaging job and the `.bintray` file.

**Comment by Zimmi48** [comment]
TODO as part of this PR: give a code owner to `dev/build/osx/make-macos-dmg.sh`

File: .gitlab-ci.yml
Comment (ejgallego) [inline .gitlab-ci.yml]: move this stuff to a `.sh` file please.

File: dev/ci/README.md
Comment (ejgallego) [inline dev/ci/README.md]: conflict marker.

**Comment by maximedenes** [comment]
@ejgallego this PR is not ready for review.

**Comment by Zimmi48** [comment]
The problem is that your pushing automatically requested a review from @ejgallego ^^

**Comment by maximedenes** [comment]
> The problem is that your pushing automatically requested a review from @ejgallego ^^

Oh, I see! Sorry, @ejgallego :)

File: .gitlab-ci.yml
Comment (Zimmi48) [inline .gitlab-ci.yml]: Remove this `after_script`. This caused an infinite loop of @coqbot retrying the failed job. The job of `after_script` is not how you use it anyways.

Comment (maximedenes) [reply .gitlab-ci.yml]: What should I put instead?

Comment (Zimmi48) [reply .gitlab-ci.yml]: You should put this in the main script. `after_script` is something that is executed *even if the build failed*.

Comment (maximedenes) [reply .gitlab-ci.yml]: > You should put this in the main script.

Ok, but how do I do that?

Comment (SkySkimmer) [reply .gitlab-ci.yml]: I'm not sure that reusing build-template is the right thing to do here.

Comment (Zimmi48) [reply .gitlab-ci.yml]: So just overwrite the default `script`.

Comment (SkySkimmer) [reply .gitlab-ci.yml]: >This caused an infinite loop of @coqbot

We should have a retry limit to avoid that/

**Review by ejgallego — CHANGES_REQUESTED** [review]

File: .gitlab-ci.yml
Comment (Zimmi48) [reply .gitlab-ci.yml]: I have looked into this and I haven't found any way of enforcing that. But this should be a really rare occurrence I think (can only happen if someone is overriding `after_script` in a PR and can be stopped by cancelling the build).

File: .gitlab-ci.yml
Comment (Zimmi48) [reply .gitlab-ci.yml]: :+1: or at the minimum 4.05 to be consistent with the current setup for Windows packages.

Comment (ejgallego) [reply .gitlab-ci.yml]: Windows packages use 4.06.1

Comment (maximedenes) [reply .gitlab-ci.yml]: 4.07 cannot be used because camlp4 is not available for that version under OPAM. So I'm trying 4.06.1.

Comment (ejgallego) [reply .gitlab-ci.yml]: I am not sure camlp4 will be ever available again for newer OCaml versions, it has been declared unmaintained. Why does the OSX build need it while neither the Linux or Windows setup does?

Comment (ejgallego) [reply .gitlab-ci.yml]: You are installing the wrong packages, in particular `lablgtk-extras`.

Comment (SkySkimmer) [reply .gitlab-ci.yml]: >4.07 cannot be used because camlp4 is not available for that version under OPAM.

opam might install different things depending on the ocaml version used, don't worry about it unless it complains about unavailable packages. If it does it should tell you why it wants the package.

Comment (ejgallego) [reply .gitlab-ci.yml]: Which also raises a question about duplication; I wonder if you could get the variables from the linux Docker image.

Comment (maximedenes) [reply .gitlab-ci.yml]: > You are installing the wrong packages, in particular lablgtk-extras.

Thanks, I'll remove this one and see.

> Which also raises a question about duplication; I wonder if you could get the variables from the linux Docker image.

The Linux Docker image, I don't think so, but booting from a macOS docker image is certainly worth trying. I'd like to do it as a second step, though, since it will probably require tweaking the tool we use to make the `.app` bundle relocatable (macOS dark magic inside).

Comment (maximedenes) [reply .gitlab-ci.yml]: > I wonder if you could get the variables from the linux Docker image.

Now I realize I misinterpreted your comment. Indeed in a setup with two Docker images, it would be interesting to know how to share variables. Maybe that's trivial with Docker?

Comment (ejgallego) [reply .gitlab-ci.yml]: > Thanks, I'll remove this one and see.

You need to add `conf-gtksourceview.2` tho, see the dependencies.

Comment (ejgallego) [reply .gitlab-ci.yml]: > Maybe that's trivial with Docker?

Certainly Docker does provide a way to have variables in Dockerfiles; indeed with our current image build setup is possible to use it.

Comment (ejgallego) [reply .gitlab-ci.yml]: It was not used originally as we were using `hub.docker` to build images, and in that context variable instantiation was more problematic.

Comment (Zimmi48) [reply .gitlab-ci.yml]: I believe that the Windows packaging job still builds camlp4 (and does not use OPAM BTW, only the test-suite Windows test does).

Comment (ejgallego) [reply .gitlab-ci.yml]: > Btw, does OPAM 2.0 come with official Windows support?

No.

**Comment by Zimmi48** [comment]
> > I believe that the Windows packaging job still builds camlp4
>
> I think this is not correct.

Indeed, you are right:

https://github.com/coq/coq/blob/bd0a681350b1bc8947d6d7603dc6a9759f0c7897/dev/build/windows/makecoq_mingw.sh#L963

But this code has not been cleaned:

https://github.com/coq/coq/blob/bd0a681350b1bc8947d6d7603dc6a9759f0c7897/dev/build/windows/makecoq_mingw.sh#L910-L928

**Comment by Zimmi48** [comment]
@maximedenes IIUC this PR is blocking on the automatic signing ability, which you have tried to implement without success. I suggest you finish the PR without the signing part, finish the related documentation PR, and open an issue with a "help wanted" label to explain the current limitations. Better do small steps than no progress at all.

**Comment by maximedenes** [comment]
> IIUC this PR is blocking on the automatic signing ability, which you have tried to implement without success. I suggest you finish the PR without the signing part, finish the related documentation PR, and open an issue with a "help wanted" label to explain the current limitations. Better do small steps than no progress at all.

But then, what is the point of moving it to gitlab, if not for signing?

**Comment by Zimmi48** [comment]
- GitLab CI artifact system means that one can fetch the resulting package even for a PR, or a previous build: it is less limited than the system put in place with Travis and bintray.

- I was thinking that, committing already this part, it would make it easier later on to get back to the signing aspect, or to request help from someone else (less things mixed up together).

Of course, these are just small considerations. If you think that in the absence of automatic signing, the drawbacks outweigh the benefits, no need to do it.

**Comment by Zimmi48** [comment]
- On GitLab CI, the timeout is longer so we could put back the test-suite and the packaging in just one job (and save overall build time).

**Comment by SkySkimmer** [comment]
So how's this going?

**Review by Zimmi48 — COMMENTED** [review]
Does the test-suite run on macOS on GitLab CI with this setup? If yes, please remove the two macOS jobs from Travis (otherwise, just the packaging one) and the `.bintray` file.

**Review by ejgallego — DISMISSED** [review]
Please provide instructions detailing how to setup a secure OSX runner setup.

**Comment by vbgl** [comment]
I’ve asked a similar question to the Inria CI support (since they are the ones providing us with macOS runners). Waiting for their answer.

**Comment by maximedenes** [comment]
> I’ve asked a similar question to the Inria CI support (since they are the ones providing us with macOS runners). Waiting for their answer.

As far as I know, Inria CI team does not provide support for Docker. They may help if they are nice today, but that's about it.

IIUC, @vbgl's attempt to use Docker on the Inria CloudStack VM failed badly (@vbgl, can you recall the error message you got?).

@ejgallego what other approaches than Docker would you call a "secure OSX runner setup"?

**Comment by ejgallego** [comment]
> @ejgallego what other approaches than Docker would you call a "secure OSX runner setup"?

I guess any setup that starts the build from a good-known image would work.

**Comment by ejgallego** [comment]
> I guess any setup that starts the build from a good-known image would work.

Ideally we would be able to setup the image using a script, but we can live with some manual work if we have no other choice.

**Comment by ejgallego** [comment]
So the setup would be:
- create a VirtualBox OSX image [many tutorials around] with the right env for Coq buildling.
- install gitlab-runner in a reasonable Linux server [the CPU should have reasonable Virt Caps]
- select the virtualbox runner https://docs.gitlab.com/runner/executors/virtualbox.html
- ?
- profit !

**Comment by maximedenes** [comment]
AFAIK, it is against the Apple EULA to run macOS on non-Apple hardware, whether directly or via a virtual machine.

**Comment by ejgallego** [comment]
For windows we can do much better as all that we need is a reasonable windows server + a windows docker image that can be built in the standard way.

**Comment by ejgallego** [comment]
> AFAIK, it is against the Apple EULA to run macOS on non-Apple hardware, whether directly or via a virtual machine.

Well then you should procure an Apple server and do the exact same steps.

**Comment by maximedenes** [comment]
> AFAIK, it is against the Apple EULA to run macOS on non-Apple hardware, whether directly or via a virtual machine.

Maybe that's no longer true, or we don't care about EULA (as long as we abide by copyright law), I have no idea. But we'd probably need confirmation.

**Comment by maximedenes** [comment]
> Well then you should procure an Apple server and do the exact same steps.

Indeed, it doesn't invalidate the whole plan.

**Comment by ejgallego** [comment]
Anyways I think the TL DR version of this is: "it is only possible to run reproducible OSX builds if we virtualize", that's the only future-proof solution.

If INRIA cannot allow us to manage the VMs, then we could try to locate a Kubernetes OSX provider, given that we are talking about 1 build per commit the pricing should be reasonable [we can keep the PR testing on Travis].

**Comment by ejgallego** [comment]
Man, OSX is in the stone age regarding cloud computing.

**Comment by SkySkimmer** [comment]
Does brew install opam 1 or 2?

**Comment by Zimmi48** [comment]
If it isn't pinned with the trick used in Travis, it installs opam 2.

**Comment by SkySkimmer** [comment]
TBH I haven't followed this PR closely, @ejgallego can you take over as assignee?

**Comment by ejgallego** [comment]
Sure.

File: .gitlab-ci.yml
Comment (SkySkimmer) [reply .gitlab-ci.yml]: Remove this retry (or explain why the bot doesn't handle it)

**Review by SkySkimmer — CHANGES_REQUESTED** [review]
I think this breaks the OSX deploy job even worse than it is now since it changes the make-dmg.sh. The travis deploy job should thus be removed.

Perhaps the build step in the make-dmg.sh should be moved to a new .sh script too (gitlab-osx-build.sh or some such)

**Comment by SkySkimmer** [comment]
How is this intended to be used? Are you going to set MACOS for everyone or just a one off job?

**Comment by Zimmi48** [comment]
So, was this version tested with `MACOS=enabled`?

**Comment by vbgl** [comment]
Yes, my gitlab mirror has this variable set.

**Comment by vbgl** [comment]
Travis job removed.

**Comment by ejgallego** [comment]
Well, Travis is quite useful I think as it doesn't require any strange hacks, the only change to fix the job there was to add a build stage, so maybe that would be worth the effort.

Also, things could work similarly for Windows.

**Comment by Zimmi48** [comment]
> Well, Travis is quite useful I think as it doesn't require any strange hacks, the only change to fix the job there was to add a build stage, so maybe that would be worth the effort.

I didn't like receiving e-mails of Travis failures every time I merged a PR for a month. The goal of this PR is not to fix Travis but to provide an alternative. If you want Travis to be kept, it should at least be commented out in `.travis.yml` until it is fixed. But I'm also not found of the Travis packaging job for another reason: the lack of artifact support means we have to workaround that and use bintray. So I'd rather remove it in favor of GitLab CI (and possibly Azure too: it could make sense duplicating the unsigned Windows / macOS packaging jobs on Azure in the future so that other people can run them on their forks). @vbgl Note that if the Travis support is removed, the bintray file should be removed too.

**Comment by ejgallego** [comment]
> So I'd rather remove it in favor of GitLab CI

Well, note that GitLab CI is not a complete replacement for Travis, as they don't provide OSX and Windows workers.

It seems to me that this is way more critical that the lack of artifact support; having to maintain the workers ourselves is really a lot of work [and in fact we are not able to provide a secure option yet]

**Comment by vbgl** [comment]
> Note that if the Travis support is removed, the bintray file should be removed too.

Can you please elaborate? Shall I do that as part of this PR? Thanks.

**Review by ejgallego — COMMENTED** [review]
Indeed as much I love Gitlab, I think that it is a pretty bad strategy for us to move Windows / OSX jobs there, as long as they don't provide workers on that platform.

After a lot of thinking it is clear to me that the Coq organization doesn't currently have the resources to manage / rent our own Windows / OSX workers; thus, I strongly advice to select another service with native OSX / Windows support.

This is a strong "don't do that" from me, but I won't block this if people want to go this way; but indeed I will soon request disabling the use of any custom worker on the main repository [thus only personal branches will be able to use this]

**Comment by vbgl** [comment]
I think I understand Emilio’s position and generally agree with him. I’d like to highlight a bit the current situation: the Travis job is broken (hence not useful); it’s very tedious to manually sign the osx package. This PR addresses both concerns, but not other concerns (as secure daily CI on OSX).

**Comment by SkySkimmer** [comment]
>If you want Travis to be kept, it should at least be commented

We're using git, no need to keep commented code around.

I think we definitely want to move windows and osx to somewhere with free workers (azure being a strong candidate so far).

Maybe we should make this PR 8.9 only to be able to get the release done and look into azure/whatever more leasurely for 8.10?

**Comment by SkySkimmer** [comment]
>Can you please elaborate? Shall I do that as part of this PR? Thanks.

yes

**Comment by vbgl** [comment]
> yes

Thanks !! I’ve just found and erased the bintray file.

**Comment by Zimmi48** [comment]
I would support making this a 8.9-only PR if it helps convincing people that this is only a temporary solution. We have complained about Travis's lack of artifact support (and short timeouts) for a very long time. I agree with everyone else that it is important to be able to run these tests on a platform with free workers. However, now that Azure is around, it looks like a strong candidate to replace both Travis and AppVeyor.

Since I was not clear in my previous comment, let me state again my current view (I think it is well aligned with what others have expressed): given all of the constraints that we have, it looks like a good trade-off would be to have packaging jobs (and probably test-suite too) for both Windows and macOS run on Azure (always), and the packaging jobs be duplicated on GitLab CI but run only on protected branches or with manually triggered pipelines (to address the security concerns that were raised before). This could possibly be implemented by having a special fork where the workers are set up and where protected branches are mirrored, and to which we push whenever we need to do some more testing. (In particular, it is unlikely that we will be able to run the Windows packaging jobs with all addons on Azure.)

**Comment by SkySkimmer** [comment]
Is there a problem with running the packaging on azure? I didn't actually check but I'm sure they've got some secure variable system.

**Comment by ejgallego** [comment]
Does Azure have OSX support?

**Review by SkySkimmer — APPROVED** [review]
If we're making this 8.9-only I guess the assignee should change to @silene

**Comment by vbgl** [comment]
Should I rebase this PR on the `v8.9` branch or open a new one?

**Comment by SkySkimmer** [comment]
Rebase, what's the point of opening a new one?

**Comment by silene** [comment]
This does not work:

https://gitlab.com/coq/coq/-/jobs/137695254

**Comment by ejgallego** [comment]
Note that merging on master would maybe be more useful. Only thing is that the main instance won't enable the CI variable.

**Comment by vbgl** [comment]
After cleaning a bit the machine on which the job is run, it did work: https://gitlab.com/coq/coq/-/jobs/137724604

**Comment by Zimmi48** [comment]
I had a quick look at the Azure doc. It looks like package signing on macOS is both supported and documented. Hopefully, this could mean having a less hackish process, but also following standard practices (which is better from a security point of view), and of course, relying less on custom infrastructure.

https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/build/xcode?view=vsts

https://docs.microsoft.com/en-us/azure/devops/pipelines/release/define-multistage-release-process?view=vsts#add-approvals

**Pull request merged**
